From: Daniel Lenski <dlenski@amazon.com>
Date: Wed, 14 Jun 2023 11:38:19 -0700
Subject: [MDEV-34009] Introduce server-initiated instant failover mechanism
 (and TLS fixes)

This is the client counterpart of MariaDB/server#3224, which provides an
initial server-initiated mechanism for instantly redirecting new client
connections to a failover server.

It modifies the Connector/C library, so that if the server responds to
our initial connection with an error packet having error number
ER_INSTANT_FAILOVER), and an error string formatted as |Human-readable
message|host[:port], the client will redirect its connection
accordingly.

This redirection is performed in mthd_my_real_connect, which appears to
be the most appropriate location given that it already contains code for
retrying/repeating a new server connection.

All new code of the whole pull request, including one or several files
that are either new files or modified ones, are contributed under the
BSD-new license. I am contributing on behalf of my employer Amazon Web
Services, Inc.

Forwarded: https://github.com/mariadb-corporation/mariadb-connector-c/pull/245
Origin: https://patch-diff.githubusercontent.com/raw/mariadb-corporation/mariadb-connector-c/pull/245.patch
Bug: https://jira.mariadb.org/browse/MDEV-15935
---
 libmariadb/include/mysql.h          |  9 +++++
 libmariadb/include/mysqld_error.h   |  3 +-
 libmariadb/libmariadb/mariadb_lib.c | 80 ++++++++++++++++++++++++++++++++++++-
 libmariadb/plugins/auth/my_auth.c   | 34 ++++++++--------
 4 files changed, 106 insertions(+), 20 deletions(-)

diff --git a/libmariadb/include/mysql.h b/libmariadb/include/mysql.h
index fdf6d1b..95e1886 100644
--- a/libmariadb/include/mysql.h
+++ b/libmariadb/include/mysql.h
@@ -228,6 +228,8 @@ extern const char *SQLSTATE_UNKNOWN;
     /* MariaDB specific */
     MYSQL_PROGRESS_CALLBACK=5999,
     MYSQL_OPT_NONBLOCK,
+    MARIADB_OPT_FOLLOW_INSTANT_FAILOVERS,
+
     /* MariaDB Connector/C specific */
     MYSQL_DATABASE_DRIVER=7000,
     MARIADB_OPT_SSL_FP,             /* deprecated, use MARIADB_OPT_TLS_PEER_FP instead */
@@ -342,12 +344,19 @@ struct st_mysql_options {
     char *bind_address;
     my_bool secure_auth;
     my_bool report_data_truncation;
+    my_bool follow_instant_failovers;
+
     /* function pointers for local infile support */
     int (*local_infile_init)(void **, const char *, void *);
     int (*local_infile_read)(void *, char *, unsigned int);
     void (*local_infile_end)(void *);
     int (*local_infile_error)(void *, char *, unsigned int);
     void *local_infile_userdata;
+
+    /* WHY are some options relegated to this "extension" structure?
+     * What is the logic for distinguishing the "main" options from
+     * the extension options?
+     */
     struct st_mysql_options_extension *extension;
 };
 
diff --git a/libmariadb/include/mysqld_error.h b/libmariadb/include/mysqld_error.h
index 1336543..ed34023 100644
--- a/libmariadb/include/mysqld_error.h
+++ b/libmariadb/include/mysqld_error.h
@@ -1236,5 +1236,6 @@
 #define ER_JSON_HISTOGRAM_PARSE_FAILED 4186
 #define ER_SF_OUT_INOUT_ARG_NOT_ALLOWED 4187
 #define ER_INCONSISTENT_SLAVE_TEMP_TABLE 4188
-#define ER_ERROR_LAST 4188
+#define ER_INSTANT_FAILOVER 4200
+#define ER_ERROR_LAST 4200
 #endif /* ER_ERROR_FIRST */
diff --git a/libmariadb/libmariadb/mariadb_lib.c b/libmariadb/libmariadb/mariadb_lib.c
index 92b79aa..288a499 100644
--- a/libmariadb/libmariadb/mariadb_lib.c
+++ b/libmariadb/libmariadb/mariadb_lib.c
@@ -84,6 +84,7 @@
 #define MA_RPL_VERSION_HACK "5.5.5-"
 
 #define CHARSET_NAME_LEN 64
+#define INSTANT_FAILOVER_LIMIT 8
 
 #undef max_allowed_packet
 #undef net_buffer_length
@@ -667,6 +668,7 @@ struct st_default_options mariadb_defaults[] =
   {{MYSQL_SET_CHARSET_NAME}, MARIADB_OPTION_STR, "default-character-set"},
   {{MARIADB_OPT_INTERACTIVE}, MARIADB_OPTION_NONE, "interactive-timeout"},
   {{MYSQL_OPT_CONNECT_TIMEOUT}, MARIADB_OPTION_INT, "connect-timeout"},
+  {{MARIADB_OPT_FOLLOW_INSTANT_FAILOVERS}, MARIADB_OPTION_BOOL, "follow-instant-failovers"},
   {{MYSQL_OPT_LOCAL_INFILE}, MARIADB_OPTION_BOOL, "local-infile"},
   {{0}, 0 ,"disable-local-infile",},
   {{MYSQL_OPT_SSL_CIPHER}, MARIADB_OPTION_STR, "ssl-cipher"},
@@ -1307,6 +1309,7 @@ mysql_init(MYSQL *mysql)
     goto error;
   mysql->options.report_data_truncation= 1;
   mysql->options.connect_timeout=CONNECT_TIMEOUT;
+  mysql->options.follow_instant_failovers= TRUE;
   mysql->charset= mysql_find_charset_name(MARIADB_DEFAULT_CHARSET);
   mysql->methods= &MARIADB_DEFAULT_METHODS;
   strcpy(mysql->net.sqlstate, "00000");
@@ -1603,7 +1606,7 @@ MYSQL *mthd_my_real_connect(MYSQL *mysql, const char *host, const char *user,
   my_bool is_multi= 0;
   char *host_copy= NULL;
   struct st_host *host_list= NULL;
-  int connect_attempts= 0;
+  int connect_attempts= 0, instant_failovers=0;
 
   if (!mysql->methods)
     mysql->methods= &MARIADB_DEFAULT_METHODS;
@@ -1743,6 +1746,7 @@ MYSQL *mthd_my_real_connect(MYSQL *mysql, const char *host, const char *user,
   else
 #endif
   {
+    tcp_redirect:
     cinfo.unix_socket=0;				/* This is not used */
     if (!port)
       port=mysql_port;
@@ -1955,6 +1959,27 @@ MYSQL *mthd_my_real_connect(MYSQL *mysql, const char *host, const char *user,
     }
   }
 
+  /* We now know the server's capabilities. If the client wants TLS/SSL,
+   * but the server doesn't support it, we should immediately abort.
+   */
+  if (mysql->options.use_ssl && !(mysql->server_capabilities & CLIENT_SSL)
+      /* Unix socket, shared memory, and named pipe are considered secure
+       * even if client requires SSL and server doesn't support it,
+       * because they do not involve off-host network traffic.
+       */
+#ifndef _WIN32
+      && net->pvio->type != PVIO_TYPE_UNIXSOCKET
+      && net->pvio->type != PVIO_TYPE_SHAREDMEM
+#else
+      && net->pvio->type != PVIO_TYPE_NAMEDPIPE
+#endif
+  )
+  {
+    SET_CLIENT_ERROR(mysql, CR_SSL_CONNECTION_ERROR, SQLSTATE_UNKNOWN,
+                     "Client requires TLS/SSL, but the server does not support it");
+    goto error;
+  }
+
   /* Set character set */
   if (mysql->options.charset_name)
     mysql->charset= mysql_find_charset_name(mysql->options.charset_name);
@@ -1973,9 +1998,59 @@ MYSQL *mthd_my_real_connect(MYSQL *mysql, const char *host, const char *user,
 
   mysql->client_flag= client_flag;
 
+  /* Until run_plugin_auth has completed, the connection
+   * cannot have been secured with TLS/SSL.
+   *
+   * This means that any client which expects to use a
+   * TLS/SSL-secured connection SHOULD NOT trust any
+   * communication received from the server prior to this
+   * point as being genuine; nor should either the client
+   * or the server send any confidential information up
+   * to this point.
+   */
+
   if (run_plugin_auth(mysql, scramble_data, scramble_len,
                              scramble_plugin, db))
+  {
+    if (mysql->net.last_errno == ER_INSTANT_FAILOVER)
+    {
+      if (!mysql->options.follow_instant_failovers)
+      {
+        /* Client has disabled instant failover. Fall through and treat this
+         * as a "normal" error. */
+      }
+      else if (instant_failovers >= INSTANT_FAILOVER_LIMIT)
+      {
+        /* Too many instant failovers */
+        my_set_error(mysql, ER_INSTANT_FAILOVER, SQLSTATE_UNKNOWN,
+                     "Too many instant failovers (>= %d)",
+                     INSTANT_FAILOVER_LIMIT);
+      }
+      else
+      {
+        char *p= mysql->net.last_error; /* Should look like '|message|host[:port]' */
+        if (p && p[0] == '|')
+          p= strchr(p + 1, '|') ? : NULL;
+        if (p && *++p) {
+          host= p;
+          p= strchr(p, ':') ? : NULL;
+          if (p) {
+            *p++ = '\0';
+            port= atoi(p);
+          }
+          else
+          {
+            /* Restore to the default port, rather than reusing our current one */
+            port= 0;
+          }
+        fprintf(stderr, "Got instant failover to '%s' (port %d)\n", host, port);
+          ++instant_failovers;
+          goto tcp_redirect;
+        }
+      }
+    }
     goto error;
+  }
 
   if (mysql->client_flag & CLIENT_COMPRESS ||
       mysql->client_flag & CLIENT_ZSTD_COMPRESSION)
@@ -3491,6 +3566,9 @@ mysql_optionsv(MYSQL *mysql,enum mysql_option option, ...)
   case MYSQL_OPT_RECONNECT:
     mysql->options.reconnect= *(my_bool *)arg1;
     break;
+  case MARIADB_OPT_FOLLOW_INSTANT_FAILOVERS:
+    mysql->options.follow_instant_failovers= *(my_bool *)arg1;
+    break;
   case MYSQL_OPT_PROTOCOL:
     mysql->options.protocol= *((uint *)arg1);
     break;
diff --git a/libmariadb/plugins/auth/my_auth.c b/libmariadb/plugins/auth/my_auth.c
index febb337..8e3efbe 100644
--- a/libmariadb/plugins/auth/my_auth.c
+++ b/libmariadb/plugins/auth/my_auth.c
@@ -316,22 +316,6 @@ static int send_client_reply_packet(MCPVIO_EXT *mpvio,
     mysql->options.extension->tls_allow_invalid_server_cert= 1;
   }
 
-  /* if server doesn't support SSL and verification of server certificate
-     was set to mandatory, we need to return an error */
-  if (mysql->options.use_ssl && !(mysql->server_capabilities & CLIENT_SSL))
-  {
-    if (!mysql->options.extension->tls_allow_invalid_server_cert ||
-        mysql->options.extension->tls_fp ||
-        mysql->options.extension->tls_fp_list)
-    {
-      my_set_error(mysql, CR_SSL_CONNECTION_ERROR, SQLSTATE_UNKNOWN,
-                          ER(CR_SSL_CONNECTION_ERROR), 
-                          "SSL is required, but the server does not support it");
-      goto error;
-    }
-  }
-
-
   /* Remove options that server doesn't support */
   mysql->client_flag= mysql->client_flag &
                        (~(CLIENT_COMPRESS | CLIENT_ZSTD_COMPRESSION | CLIENT_SSL | CLIENT_PROTOCOL_41) 
@@ -411,8 +395,19 @@ static int send_client_reply_packet(MCPVIO_EXT *mpvio,
   {
     unsigned int verify_flags= 0;
     /*
-      Send mysql->client_flag, max_packet_size - unencrypted otherwise
-      the server does not know we want to do SSL
+      Send UNENCRYPTED "Login Request" packet with mysql->client_flag
+      and max_packet_size, but no username; without this, the server
+      does not know we want to switch to SSL/TLS
+
+      FIXME: Sending this packet is a very very VERY bad idea. It
+      contains the client's preferred charset and flags in plaintext;
+      this can be used for fingerprinting the client software version,
+      and probable geographic location.
+
+      This offers a glaring opportunity for pervasive attackers to
+      easily target, intercept, and exploit the client-server
+      connection (e.g. "MITM all connections from known-vulnerable
+      client versions originating from countries X, Y, and Z").
     */
     if (ma_net_write(net, (unsigned char *)buff, (size_t) (end-buff)) || ma_net_flush(net))
     {
@@ -422,6 +417,9 @@ static int send_client_reply_packet(MCPVIO_EXT *mpvio,
                           errno);
       goto error;
     }
+    /* This is where the socket is actually converted from a plain
+     * TCP/IP socket to a TLS/SSL-wrapped socket.
+     */
     mysql->net.tls_verify_status = 0;
     if (ma_pvio_start_ssl(mysql->net.pvio))
       goto error;
